### DISCUSSION TITLE
Incresed Observability of go-redis OSS client

### DISCUSSION THREAD

**Discussion by srikar-jilugu** [redis/go-redis#3097] [Ideas]
Currently, our OSS Redis Cluster ecosystem has limited monitoring capabilities within the go-redis client, making it challenging to debug issues from the client's perspective. For instance, when there's a sudden increase in redis response time during an Elasticache autoscaling event,, we want to check if there are any issues due to prolonged `moved` errors, an excessive number of retries, or if the cluster topology hasn't been updated on the client side

Along with that, we recently enabled `RoutebyLatency` in one of our cluster clients and there is very little scope to validate its functionality as there's no insight into which node is handling specific requests from the client.

To enhance the observability of the client, we could implement the following improvements through metrics or logging, depending on the use case:

- Error logs for `moved` and `readonly` errors 
- Metric for `no.of.retries` per command
- Metrics/ Debug logging to know whenever a node is marked as `failed`
- A way to know to which nodes are receiving requests from a specific redis client(a pod/container) are going, could be a metric with the node ip as label

**Comment by srikar-jilugu** [comment]
@monkey92t @vmihailenco please take a look at this when you are free

**Comment by srikar-jilugu** [comment]
@monkey92t Could you take a look at this? I’d appreciate any feedback or suggestions you might have. I'm happy to submit a proposal PR if needed.

**Reply by monkey92t** [reply]
> @monkey92t Could you take a look at this? I’d appreciate any feedback or suggestions you might have. I'm happy to submit a proposal PR if needed.

Currently, go-redis is maintained by the Redis organization.😊

**Reply by srikar-jilugu** [reply]
@monkey92t oh cool, thanks for letting me know!

**Comment by srikar-jilugu** [comment]
@vladvildanov Could you please review this idea? I'd love to hear your thoughts and feedback on it.

**Comment by vladvildanov** [comment]
@srikar-jilugu So it's a matter of the performance here. I would say we cannot just do "some" metrics or logging, it should be a strategy of what we want an achieve and what does it costs in terms of performance. I would appreciate any thoughts on this 🙂
